subset/SSet.cpp

```diff
@@ -0,0 +1,105 @@
+#include "SSet.h"
+#include <algorithm>
+#include <iterator>
+
+bool SSet::checkSumOptimized(DataType expected) {
+  positive_sum = 0;
+  negative_sum = 0;
+  for (auto n : set_) {
+    if (n > 0) positive_sum += n;
+    if (n < 0) negative_sum += n;
+  }
+  return set_.empty() ? false
+                      : checkSumOptimizedRecursive(set_.size() - 1, expected);
+}
+
+bool SSet::checkSumOptimizedRecursive(int i, DataType expected) {
+  if (i == 0) return set_[i] == expected;
+  if ((expected > positive_sum) || (expected < negative_sum)) return false;
+  return checkSumOptimizedRecursive(i - 1, expected) || (set_[i] == expected) ||
+         checkSumOptimizedRecursive(i - 1, expected - set_[i]);
+}
+
+std::vector<SSet::DataType> SSet::subsetSums(std::vector<DataType>* set_p) {
+  std::vector<DataType> set = *set_p;
+  long total = 1 << set.size();  // total number of subsets = size of power set = 2^n
+  std::vector<SSet::DataType> sums(total, 0);
+  sums[1] = set[0];
+  int effectiveBits = 1, prevPowOf2 = 1;
+  for (long long i = 2; i < total; ++i) {
```

Comment:
There is a decorrelation with `long` and `long long` types used. I suggest to use std::size_t or std::uint64_t

Reply:
I think the same logic can be applied to `effectiveBits` and `prevPowOf2`